### DISCUSSION TITLE
How to cache signed user? Avalonia/MSAL

### DISCUSSION THREAD

**Discussion by DmitryBorodiy** [microsoftgraph/msgraph-sdk-dotnet#2540] [Q&A] [answered]
```
using Azure.Identity;
using Microsoft.Graph;
using System.Threading.Tasks;

namespace QuickFind.Services.Account
{
    public static class MsalUserProvider
    {
        #region Consts
        private const string MsalClientId = "ee1e4fbf-0998-4df2-92ed-dff3f527a69c";
        private const string MsalTenantId = "98c891c3-0990-45af-bd67-c07e182a14e5";
        #endregion

        #region Props
        public static GraphServiceClient UserClient { get; set; }
        private static InteractiveBrowserCredential _deviceCodeCredential;
        private static InteractiveBrowserCredential _interactiveBrowserCredential;
        private const string CacheFile = "Properties/temp.bin";

        private static string[] Scopes => 
        [
            "User.Read",
            "User.Read.All",
            "Mail.ReadBasic",
            "Mail.Read",
            "Tasks.Read",
            "Tasks.ReadWrite",
            "Tasks.Read.Shared",
            "Tasks.ReadWrite.Shared",
        ];
        #endregion

        #region Methods

        public static async Task InitializeGraphForUserAuth()
        {
            var options = new InteractiveBrowserCredentialOptions() { TokenCachePersistenceOptions = new TokenCachePersistenceOptions() { Name = "QuickFind.Temp" } };
            _interactiveBrowserCredential = new InteractiveBrowserCredential(MsalTenantId, MsalClientId, options);

            UserClient = new GraphServiceClient(_interactiveBrowserCredential, Scopes);

            if(UserClient != null)
            {
                var user = await UserClient.Me.GetAsync();

                App.CurrentUser = new UserInformation();
                App.CurrentUser.InitFromGraph(user);
            }
        }

        public static async void InitializeGraph()
        {
            await InitializeGraphForUserAuth();
        }

        #endregion
    }
}
```

Hello, Microsoft Graph community!
How to enable cashing for InteractiveBrowserCredential?

**Comment by andrueastman** [comment]
Thanks for raising this @DmitryBorodiy 

Any chance if the documentation at the link below is useful for your scenario? https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/identity/Azure.Identity/samples/TokenCache.md

**Reply by DmitryBorodiy** [reply]
I already seem to be using the method that is specified in the documentation, but it just doesn't work.

**Reply by DmitryBorodiy** [reply]
Moreover, there is another problem, although I successfully log into my account and create a GraphServiceClient, for some reason it returns everything empty, I can’t get anything except the login and email, I’m trying to get something like todo, as a result I get an ODataError  with ItemNotFound error.

**Reply by andrueastman** [reply]
Any chance you can share a sample of how you are making the requests? Typically `ItemNotFound` means that you're token is ok but the data doesn't exist. Otherwise, you would ideally get an `unauthorized` or `unauthenticated` error in a scenario where the token is the issue.

**Reply by DmitryBorodiy** [reply]
Regarding the caching of the JWT token. I just get the token using the GetToken() method and write it to the application's temporary files.

**Reply by DmitryBorodiy** [reply]
Regarding ODataError, I suspect I didn't specify scopes correctly, can you tell if there are bugs in my code. Perhaps scopes should be written with small letters.

**Reply by DmitryBorodiy** [reply, marked as answer]
> Any chance you can share a sample of how you are making the requests? Typically `ItemNotFound` means that you're token is ok but the data doesn't exist. Otherwise, you would ideally get an `unauthorized` or `unauthenticated` error in a scenario where the token is the issue. 

Hey, I found problem! The problem is that I specified the tenant id. I removed it, and now everything worked.

**Reply by DmitryBorodiy** [reply]
I also use PublicClientApplicationBuilder class instead of InteractiveBrowserCredential and initialize GraphServiceClient from HttpClient with access token authorization header.
